### ISSUE TITLE
aws_route53_record should create records for aws_instance resource

### ISSUE THREAD

**Issue by hgfranco** [hashicorp/terraform#19269]
### Terraform Version
```
terraform -v
Terraform v0.11.10
+ provider.aws v1.41.0
```

### Terraform Configuration Files
This works:
```
// Create a CNAME for the ALB
resource "aws_route53_record" "route53_alb" {
  zone_id = "${var.route_53_zone_id}"
  name    = "alb_app.${var.route_53_zone}"
  type    = "CNAME"
  ttl     = "300"
  records = ["${aws_alb.my_alb.dns_name}"]
}
```

This doesn't work:
```
// Create A records for EC2 instances
resource "aws_route53_record" "route53_ec2" {
  zone_id = "${var.route_53_zone_id}"
  name    = "app0${count.index + 1}.myapp.com"
  type    = "A"
  ttl     = "300"
  records = ["${aws_instance.my_app.private_ip}"]
}
```

### Expected Behavior
What should have happened?
I should have a couple of A records created.

### Actual Behavior
What actually happened?
This error when creating records for EC2 instances:
```Resource 'aws_instance.my_app' not found for variable 'aws_instance.my_app.private_ip'```

### Steps to Reproduce
Please list the full steps required to reproduce the issue, for example:
1. `terraform plan`

**Comment by hgfranco** [comment]
It appears that none of the "attributes references" listed here seem to work.
https://www.terraform.io/docs/providers/aws/r/instance.html

**Comment by hgfranco** [comment]
@apparentlymart thanks for the follow up.  Unfortunately it doesn't appear to be the same issue.  I pulled the terraform log and couldn't find any errors besides what I listed above.  Does TF_LOG contain any sensitive info? It appears not.  I can post it here.

**Comment by apparentlymart** [comment]
Hi @hgfranco! Thanks for checking.

The log file does not contain credentials, so for a contrived test case it can be fine to share it. It will disclose some details about how your infrastructure is set up, and whether _that_ is "sensitive information" is something you'd need to decide. If you are willing to share the whole log, please post it in [a gist](https://gist.github.com/), since the logs are too long/wide to be readable directly inside a comment.

**Comment by hgfranco** [comment]
@apparentlymart here's the [gist](https://gist.github.com/hgfranco/3a6fe2e89064e983907ce6ae094046a6)

**Comment by apparentlymart** [comment]
Hi @hgfranco! Thanks for sharing that.

Since the names of things in the log don't exactly match the configuration you shared originally I had to guess a little, but I'm assuming that where you showed `aws_instance.my_app.private_ip` in your example here that'd be `aws_instance.nexus_ec2.private_ip` in this log.

One thing I notice from comparing the two is it looks like the instance resource has `count` set -- I think set to 3, if I'm reading the log correctly -- and so references to it ought to be using the splat syntax:

```hcl
// Create A records for EC2 instances
resource "aws_route53_record" "route53_ec2" {
  count = 3 # or whatever variable controls how many instances there are

  zone_id = "${var.route_53_zone_id}"
  name    = "app0${count.index + 1}.myapp.com"
  type    = "A"
  ttl     = "300"
  records = ["${aws_instance.nexus_ec2.*.private_ip[count.index]}"]
}
```

I wonder if perhaps this is a case of Terraform's validation not detecting that problem and then it later getting caught during expression evaluation, resulting in a more confusing error. Updating the configuration to be similar to what I showed above may allow you to get past this strange error.

Some good news is that a new configuration engine is in master ready to be included in the forthcoming v0.12.0 release, and not only does that have more robust validation, it will also allow this to be expressed in a more readable way:

```hcl
  records = [aws_instance.nexus_ec2[count.index].private_ip]
```
